src/block/Composter.php

```diff
@@ -0,0 +1,190 @@
+use pocketmine\math\Vector3;
+use pocketmine\network\mcpe\protocol\SpawnParticleEffectPacket;
+use pocketmine\network\mcpe\protocol\types\DimensionIds;
+use pocketmine\player\Player;
+use pocketmine\world\sound\ComposterEmptySound;
+use pocketmine\world\sound\ComposterFillSound;
+use pocketmine\world\sound\ComposterFillSuccessSound;
+use pocketmine\world\sound\ComposterReadySound;
+use function abs;
+use function mt_rand;
+
+class Composter extends Transparent{
+	public const CROP_GROWTH_EMITTER_PARTICLE = "minecraft:crop_growth_emitter";
+
+	protected int $composter_fill_level = 0;
+
+	protected function writeStateToMeta() : int{
+		return $this->composter_fill_level;
+	}
+
+	public function writeStateToItemMeta() : int{
+		return $this->composter_fill_level;
+	}
+
+	public function readStateFromData(int $id, int $stateMeta) : void{
+		$this->composter_fill_level = BlockDataSerializer::readBoundedInt("composter_fill_level", $stateMeta, 0, 8);
+	}
+
+	public function getStateBitmask() : int{
+		return 0b1111;
+	}
+
+	protected function recalculateCollisionBoxes() : array{
+		$boxes = [$this->getSideCollisionBox(Facing::DOWN)];
+		foreach (Facing::HORIZONTAL as $side) {
+			$boxes[] = $this->getSideCollisionBox($side);
+		}
+		return $boxes;
+	}
+
+	protected function getSideCollisionBox(int $face = Facing::NORTH) : AxisAlignedBB{
+		$empty = abs(15 - 2 * $this->composter_fill_level) - (int) ($this->composter_fill_level === 0);
+		return ($face === Facing::DOWN || $face === Facing::UP) ? AxisAlignedBB::one()->contract(2 / 16, 0, 2 / 16)->trim(Facing::UP, $empty / 16) : AxisAlignedBB::one()->trim(Facing::opposite($face), 14 / 16);
+	}
+
+	public function isEmpty() : bool{
+		return $this->composter_fill_level === 0;
+	}
+
+	public function isReady() : bool{
+		return $this->composter_fill_level === 8;
+	}
+
+	public function getComposterFillLevel() : int{
+		return $this->composter_fill_level;
+	}
+
+	/**
+	 * @throws \Exception
+	 */
+	public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player $player = null) : bool{
+		if ($player instanceof Player && !$player->isSneaking()) {
+			if ($this->compost($player, clone $item)) {
+				$item->pop();
+			}
+		}
+		return true;
+	}
+
+	public function pushCollidedEntities() : void{
+		if ($this->composter_fill_level === 7) return;
+		foreach (
+			$this->position->getWorld()->getNearbyEntities(
+				$this->getSideCollisionBox(Facing::DOWN)->extend(Facing::UP, 0.25)->offset(
+					$this->position->getFloorX(),
+					$this->position->getFloorY(),
+					$this->position->getFloorZ()
+				)
+			) as $entity) {
+			if ($entity instanceof Player || $entity instanceof Projectile) continue;
+			$motion = $entity->getMotion();
+			$motion->y = 0.2; //Lower can make entity clip through the block, the game cannot catch up the speed of the entity? :/
+			$entity->setMotion($motion);
+		}
+	}
+
+	protected function spawnParticleEffect() : void{
+		$packet = SpawnParticleEffectPacket::create(
```

Comment:
That's not how this is done. We don't use packets in core code.

Reply:
> That's not how this is done. We don't use packets in core code.

You can't do it without using packet. Spawn particle separately is okay but it's not the same in vanilla

Reply:
> > That's not how this is done. We don't use packets in core code.
> 
> You can't do it without using packet. Spawn particle separately is okay but it's not the same in vanilla

Obviously, packet is required. My point is that core code is not supposed to use them directly. This should be done via a Particle class.